charmtools/build/tactics.py

```diff
@@ -750,35 +751,44 @@ def combine(self, existing):
         self.previous = existing.previous + [existing]
         return self
 
-    def _add(self, pip, wheelhouse, *reqs):
+    def _add(self, wheelhouse, *reqs):
         with utils.tempdir(chdir=False) as temp_dir:
             # put in a temp dir first to ensure we track all of the files
-            utils.Process((pip, 'install',
-                           '--no-binary', ':all:',
-                           '-d', temp_dir) +
-                          reqs).exit_on_error()()
+            self._pip('download', '--no-binary', ':all:', '-d', temp_dir,
+                      *reqs)
             for wheel in temp_dir.files():
                 dest = wheelhouse / wheel.basename()
                 dest.remove_p()
                 wheel.move(wheelhouse)
                 self.tracked.append(dest)
 
+    def _pip(self, *args):
+        # have to use bash to call pip to activate the venv properly first
+        utils.Process(('bash', '-c', ' '.join(
+            ('.', self._venv / 'bin' / 'activate', ';', 'pip3') + args
```

Comment:
It's an internal method that should only be called from __call__.  I can add an assertion; it will blow up either way but maybe the error will be more useful.